src/approaches/gnn_approach.py

```diff
@@ -0,0 +1,445 @@
+        # Seed torch.
+        torch.manual_seed(self._seed)
+
+    @abc.abstractmethod
+    def _extract_target_from_data(self, segment: Segment,
+                                  segment_traj: List[Segment],
+                                  ll_traj: LowLevelTrajectory) -> _Output:
+        """Given a segment in the data, which was part of the given segment
+        trajectory associated with the given low-level trajectory, extract the
+        target output for the GNN."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _setup_output_specific_fields(
+        self, data: List[Tuple[State, Set[GroundAtom], Set[GroundAtom],
+                               _Output]]
+    ) -> None:
+        """Given the dataset of inputs and outputs, set up any necessary
+        output-specific fields."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _graphify_single_target(self, datapoint: Tuple[State, Set[GroundAtom],
+                                                       Set[GroundAtom],
+                                                       _Output],
+                                graph_input: Dict,
+                                object_to_node: Dict) -> Dict:
+        """Given a single datapoint (input and output) and the return values of
+        graphify_single_input on the input, return a target graph."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _criterion(self, output: torch.Tensor,
+                   target: torch.Tensor) -> torch.Tensor:
+        """Define the criterion function for passing into train_model()."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _global_criterion(self, output: torch.Tensor,
+                          target: torch.Tensor) -> torch.Tensor:
+        """Define the global criterion function for passing into
+        train_model()."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _add_output_specific_fields_to_save_info(self, info: Dict) -> None:
+        """Given a dict of info to be saved, add output-specific fields to
+        it."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _load_output_specific_fields_from_save_info(self, info: Dict) -> None:
+        """Given a dict of saved info, load output-specific fields from it."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _extract_output_from_graph(self, graph_output: Dict,
+                                   object_to_node: Dict) -> _Output:
+        """At evaluation time, given an output GNN, extract the actual
+        output."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
```

Comment:
i'm assuming this function is basically the same as before

Reply:
yup